tests/integ/test_vpc.py

```diff
@@ -0,0 +1,364 @@
+    pickle_args = {} if sys.version_info.major == 2 else {'encoding': 'latin1'}
+    with gzip.open(data_path, 'rb') as f:
+        test_set, _, _ = pickle.load(f, **pickle_args)
+
+    training_data_path = os.path.join(DATA_DIR, 'dummy_tensor')
+    prefix = 'test_byo_estimator'
+    key = 'recordio-pb-data'
+    s3_train_data = sagemaker_session.upload_data(path=training_data_path,
+                                                  key_prefix=os.path.join(prefix, 'train', key))
+
+    ec2_client = sagemaker_session.boto_session.client('ec2')
+    subnet_ids, security_group_id = get_or_create_subnets_and_security_group(ec2_client,
+                                                                             sagemaker_session.boto_session.region_name)
+
+    estimator = Estimator(image_name=image_name,
+                          role='SageMakerRole',
+                          train_instance_count=1,
+                          train_instance_type=instance_type,
+                          sagemaker_session=sagemaker_session,
+                          base_job_name='test-vpc-byo-fm',
+                          subnets=subnet_ids,
+                          security_group_ids=[security_group_id])
+    estimator.set_hyperparameters(num_factors=10,
+                                  feature_dim=784,
+                                  mini_batch_size=100,
+                                  predictor_type='binary_classifier')
+
+    with timeout(minutes=TRAINING_DEFAULT_TIMEOUT_MINUTES):
+        estimator.fit({'train': s3_train_data})
+        print('training job succeeded: {}'.format(estimator.latest_training_job.name))
+
+    job_desc = sagemaker_session.sagemaker_client.describe_training_job(
+        TrainingJobName=estimator.latest_training_job.name)
+    assert set(subnet_ids) == set(job_desc['VpcConfig']['Subnets'])
+    assert [security_group_id] == job_desc['VpcConfig']['SecurityGroupIds']
+
+    endpoint_name = estimator.latest_training_job.name
+    with timeout_and_delete_endpoint_by_name(endpoint_name, sagemaker_session):
+        model = estimator.create_model()
+        predictor = model.deploy(1, 'ml.m4.xlarge', endpoint_name=endpoint_name)
+        predictor.serializer = fm_serializer
+        predictor.content_type = 'application/json'
+        predictor.deserializer = json_deserializer
+
+        model_desc = sagemaker_session.sagemaker_client.describe_model(ModelName=model.name)
+        assert set(subnet_ids) == set(model_desc['VpcConfig']['Subnets'])
+        assert [security_group_id] == model_desc['VpcConfig']['SecurityGroupIds']
+
+        result = predictor.predict(test_set[0][:10])
+
+        assert len(result['predictions']) == 10
+        for prediction in result['predictions']:
+            assert prediction['score'] is not None
+
+
+def fm_serializer(data):
+    js = {'instances': []}
+    for row in data:
+        js['instances'].append({'features': row.tolist()})
+    return json.dumps(js)
```

Comment:
I appreciate the effort to write vpc tests for each framework. One concern is that each of these integration tests can take even 45 minutes, depending of the instance type. Given that all the frameworks share the same Estimator, is it possible to add vpc in some of the integration tests previously created for the frameworks to avoid repetitive tests?

Arguably vpc should be part of other integrations tests and not integration tests itself.

Reply:
I suggest to check how longer will be the test run after your changes.
If you have the same functionality replicated in multiple classes that might be a sign that some refactoring could be done.

Reply:
After more thinking about this (and witnessing first-hand that the VPC tests took a long time to run), I've decided to consolidate. VPC features will be covered by a single cases in the TensorFlow integ tests. Other functionality like changing/dropping VPC config will be added in unit tests.